lambda/debuginfo.ml

```diff
@@ -64,6 +64,17 @@ module Scoped_location = struct
     | [] -> "<unknown>"
     | scopes -> String.concat "" (to_strings [] scopes)
 
+  let string_of_scopes =
+    let module StringSet = Set.Make (String) in
```

Comment:
What about using a Weak table here?

Reply:
My intuition is that it would make builds non deterministic.

Reply:
I don't think so: you'll still have the guarantee that equal strings are physically equal.

Reply:
Sorry, I am confused: either strings may be removed from the set and some
sharing might be lost, or they are never removed and the nature of the
container does not matter.

Reply:
By definition, all data marshaled to a given file remain alive until the marshaling happen (so, they won't go out of the weak table); since they will all have gone through the hash-consing (Weak.merge), sharing of equal strings is guaranteed.

Reply:
Btw, are you confident that the calculation of string_of_scopes itself is not too costly?

Reply:
Another, more standard approach would be to just memoize the function in a mutable map. If I understand correctly, this should give the same sharing-outputs benefits (as long as we are not sharing the same non-constant result among different strings, which I don't think we are?), and it avoids recomputing the function completely.

Reply:
For a sequential build of `world.opt`, I see an increase of about 10 seconds,
the baseline being 4 minutes.

Reply:
@xclerc which version is giving this increase of 10 seconds compared to what? Are you comparing your PR to trunk, or Alain or my suggestion to your PR?

Reply:
It was with/without the diff in this PR.